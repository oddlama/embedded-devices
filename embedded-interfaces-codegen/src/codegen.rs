//! Code generation

use proc_macro2::TokenStream as TokenStream2;
use quote::{format_ident, quote};
use syn::Ident;

use crate::parser::*;

pub fn generate_registers(registers_def: &RegistersDefinition) -> syn::Result<TokenStream2> {
    let mut generated = TokenStream2::new();

    // Generate code for each register
    for register in &registers_def.registers {
        let register_code = generate_register(registers_def, register)?;
        generated.extend(register_code);
    }

    Ok(generated)
}

fn generate_register(registers_def: &RegistersDefinition, register: &RegisterDefinition) -> syn::Result<TokenStream2> {
    let register_name = &register.name;
    let unpacked_name = format_ident!("{}Unpacked", register_name);

    // Get effective attributes (defaults + register-specific)
    let attrs = registers_def.get_effective_attrs(register);

    // Extract register attributes
    let (addr, mode, size) = extract_register_attrs(&attrs)?;
    let (codec_error, i2c_codec, spi_codec) = extract_codec_attrs(&attrs)?;

    // Generate unpacked struct with non-reserved fields
    let unpacked_struct = generate_unpacked_struct(register, &unpacked_name)?;

    // Generate packed struct
    let packed_struct = generate_packed_struct(register, register_name, size)?;

    // Generate trait implementations
    let trait_impls = generate_trait_implementations(
        register_name,
        &unpacked_name,
        &addr,
        &mode,
        size,
        &codec_error,
        &i2c_codec,
        &spi_codec,
    )?;

    Ok(quote! {
        #unpacked_struct
        #packed_struct
        #trait_impls
    })
}

fn extract_register_attrs(attrs: &[RegisterAttr]) -> syn::Result<(TokenStream2, String, usize)> {
    let mut addr = None;
    let mut mode = None;
    let mut size = None;

    for attr in attrs {
        match attr.name.to_string().as_str() {
            "addr" => {
                if let RegisterAttrValue::Int(lit_int) = &attr.value {
                    addr = Some(quote! { #lit_int });
                }
            }
            "mode" => {
                if let RegisterAttrValue::Ident(ident) = &attr.value {
                    mode = Some(ident.to_string());
                }
            }
            "size" => {
                if let RegisterAttrValue::Int(lit_int) = &attr.value {
                    size = Some(lit_int.base10_parse::<usize>()?);
                }
            }
            _ => {}
        }
    }

    let addr = addr.ok_or_else(|| syn::Error::new_spanned(&attrs[0].name, "Missing addr attribute"))?;
    let mode = mode.ok_or_else(|| syn::Error::new_spanned(&attrs[0].name, "Missing mode attribute"))?;
    let size = size.ok_or_else(|| syn::Error::new_spanned(&attrs[0].name, "Missing size attribute"))?;

    Ok((addr, mode, size))
}

fn extract_codec_attrs(attrs: &[RegisterAttr]) -> syn::Result<(TokenStream2, TokenStream2, TokenStream2)> {
    let mut codec_error = None;
    let mut i2c_codec = None;
    let mut spi_codec = None;

    for attr in attrs {
        match attr.name.to_string().as_str() {
            "codec_error" => match &attr.value {
                RegisterAttrValue::Ident(ident) => {
                    codec_error = Some(quote! { #ident });
                }
                RegisterAttrValue::String(lit_str) => {
                    let ident: Ident = syn::parse_str(&lit_str.value())?;
                    codec_error = Some(quote! { #ident });
                }
                _ => {}
            },
            "i2c_codec" => match &attr.value {
                RegisterAttrValue::Ident(ident) => {
                    i2c_codec = Some(quote! { #ident });
                }
                RegisterAttrValue::String(lit_str) => {
                    let ident: Ident = syn::parse_str(&lit_str.value())?;
                    i2c_codec = Some(quote! { #ident });
                }
                _ => {}
            },
            "spi_codec" => match &attr.value {
                RegisterAttrValue::Ident(ident) => {
                    spi_codec = Some(quote! { #ident });
                }
                RegisterAttrValue::String(lit_str) => {
                    let ident: Ident = syn::parse_str(&lit_str.value())?;
                    spi_codec = Some(quote! { #ident });
                }
                _ => {}
            },
            _ => {}
        }
    }

    let codec_error = codec_error.unwrap_or_else(|| quote! { () });
    let i2c_codec = i2c_codec.unwrap_or_else(|| quote! { () });
    let spi_codec = spi_codec.unwrap_or_else(|| quote! { () });

    Ok((codec_error, i2c_codec, spi_codec))
}

fn generate_unpacked_struct(register: &RegisterDefinition, unpacked_name: &Ident) -> syn::Result<TokenStream2> {
    let mut fields = Vec::new();
    let mut default_values = Vec::new();

    // Copy documentation from register
    let doc_attrs = &register.attributes;

    for field in &register.fields {
        // Skip reserved fields (fields without names)
        if let Some(field_name) = &field.name {
            let field_type = &field.field_type;
            let field_attrs = &field.attributes;

            // Generate field documentation
            let default_doc = if let Some(default_val) = &field.default_value {
                quote! { #[doc = concat!("Default: `", stringify!(#default_val), "`")] }
            } else {
                quote! { #[doc = "Default: not specified"] }
            };

            fields.push(quote! {
                #(#field_attrs)*
                #[doc = ""]
                #default_doc
                pub #field_name: #field_type
            });

            // Collect default values
            if let Some(default_val) = &field.default_value {
                default_values.push(quote! { #field_name: #default_val });
            } else {
                default_values.push(quote! { #field_name: Default::default() });
            }
        }
    }

    Ok(quote! {
        #(#doc_attrs)*
        #[derive(Copy, Clone, PartialEq, core::fmt::Debug, defmt::Format)]
        pub struct #unpacked_name {
            #(#fields,)*
        }

        impl Default for #unpacked_name {
            fn default() -> Self {
                Self {
                    #(#default_values,)*
                }
            }
        }
    })
}

fn generate_packed_struct(
    register: &RegisterDefinition,
    register_name: &Ident,
    size: usize,
) -> syn::Result<TokenStream2> {
    let unpacked_name = format_ident!("{}Unpacked", register_name);

    Ok(quote! {
        #[doc = concat!("This is the packed representation of [`", stringify!(#register_name), "`].")]
        #[derive(Copy, Clone, PartialEq, Eq, bytemuck::Pod, bytemuck::Zeroable)]
        #[repr(transparent)]
        pub struct #register_name([u8; #size]);

        impl #register_name {
            #[inline]
            #[doc = concat!("Pack all fields in the given [`", stringify!(#unpacked_name), "`] representation.")]
            pub fn new(value: #unpacked_name) -> Self {
                Self(value.into_bytes())
            }

            #[inline]
            #[doc = concat!("Unpack all fields and return them as a [`", stringify!(#unpacked_name), "`]. If you don't need all fields, this is more expensive than just using the appropriate `read_*` functions directly.")]
            pub fn read_all(&self) -> #unpacked_name {
                #unpacked_name::from_bytes(self.0)
            }

            #[inline]
            #[doc = concat!("Pack all fields in the given [`", stringify!(#unpacked_name), "`] representation. If you only want to write some fields, this is more expensive than just using the appropriate `write_*` functions directly.")]
            pub fn write_all(&mut self, value: #unpacked_name) {
                self.0 = value.into_bytes();
            }
        }

        impl Default for #register_name {
            fn default() -> Self {
                Self(#unpacked_name::default().into_bytes())
            }
        }
    })
}

fn generate_trait_implementations(
    register_name: &Ident,
    unpacked_name: &Ident,
    addr: &TokenStream2,
    mode: &str,
    size: usize,
    codec_error: &TokenStream2,
    i2c_codec: &TokenStream2,
    spi_codec: &TokenStream2,
) -> syn::Result<TokenStream2> {
    // Generate readable/writable traits based on mode
    let mode_traits = match mode {
        "r" => quote! { impl embedded_interfaces::registers::ReadableRegister for #register_name {} },
        "w" => quote! { impl embedded_interfaces::registers::WritableRegister for #register_name {} },
        "rw" => quote! {
            impl embedded_interfaces::registers::ReadableRegister for #register_name {}
            impl embedded_interfaces::registers::WritableRegister for #register_name {}
        },
        _ => {
            return Err(syn::Error::new_spanned(
                register_name,
                format!("Invalid mode: {}", mode),
            ));
        }
    };

    Ok(quote! {
        impl embedded_interfaces::registers::Register for #register_name {
            type Unpacked = #unpacked_name;
            type CodecError = #codec_error;
            type SpiCodec = #spi_codec;
            type I2cCodec = #i2c_codec;
            const REGISTER_SIZE: usize = #size;
            const ADDRESS: u64 = #addr;
        }

        #mode_traits

        impl AsRef<#unpacked_name> for #unpacked_name {
            #[inline]
            fn as_ref(&self) -> &#unpacked_name {
                self
            }
        }

        impl AsRef<#register_name> for #register_name {
            #[inline]
            fn as_ref(&self) -> &#register_name {
                self
            }
        }

        impl From<&#register_name> for #unpacked_name {
            #[inline]
            fn from(value: &#register_name) -> Self {
                #unpacked_name::from_bytes(value.0)
            }
        }

        impl From<#register_name> for #unpacked_name {
            #[inline]
            fn from(value: #register_name) -> Self {
                #unpacked_name::from_bytes(value.0)
            }
        }

        impl From<&#unpacked_name> for #register_name {
            #[inline]
            fn from(value: &#unpacked_name) -> Self {
                Self(value.into_bytes())
            }
        }

        impl From<#unpacked_name> for #register_name {
            #[inline]
            fn from(value: #unpacked_name) -> Self {
                Self(value.into_bytes())
            }
        }
    })
}
