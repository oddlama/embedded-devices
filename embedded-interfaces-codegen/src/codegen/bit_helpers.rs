use proc_macro2::TokenStream;
use quote::quote;
use syn::{Ident, Type};

use super::bit_pattern::NormalizedRange;

/// Get the size in bits for a primitive type
pub fn get_type_bits(type_name: &str) -> usize {
    match type_name {
        "bool" => 1,
        "u8" | "i8" => 8,
        "u16" | "i16" => 16,
        "u32" | "i32" | "f32" => 32,
        "u64" | "i64" | "f64" => 64,
        "u128" | "i128" => 128,
        _ => panic!("Unknown type: {type_name}"),
    }
}

/// Get the size in bits for an element type (recursive for arrays)
pub fn get_element_bits(element_type: &Type) -> Result<usize, String> {
    match element_type {
        Type::Path(type_path) => {
            if let Some(ident) = type_path.path.get_ident() {
                let type_name = ident.to_string();
                Ok(get_type_bits(&type_name))
            } else {
                Err("Complex path types are not yet supported".to_string())
            }
        }
        Type::Array(array_type) => {
            let inner_bits = get_element_bits(&array_type.elem)?;
            if let syn::Expr::Lit(syn::ExprLit {
                lit: syn::Lit::Int(lit_int),
                ..
            }) = &array_type.len
            {
                let array_len = lit_int.base10_parse::<usize>().map_err(|e| e.to_string())?;
                Ok(inner_bits * array_len)
            } else {
                Err("Array length must be a literal integer".to_string())
            }
        }
        _ => Err("Unsupported element type for bit calculation".to_string()),
    }
}

/// Generates code that copies a range of bits from the source to the destination.
///
/// # Arguments
/// * `src_bits_ident` - The identifier for the source bitslice
/// * `dst_bits_ident` - The identifier for the destination bitslice
/// * `src_range` - The bit range in the source (start..end)
/// * `dst_range` - The bit range in the destination (start..end)
pub fn generate_bit_range_copy(
    src_bits_ident: &Ident,
    dst_bits_ident: &Ident,
    (src_start, src_end): (usize, usize),
    (dst_start, dst_end): (usize, usize),
) -> TokenStream {
    quote! {{
        #dst_bits_ident[#dst_start..#dst_end].copy_from_bitslice(&#src_bits_ident[#src_start..#src_end]);
    }}
}

pub fn generate_copy_from_normalized_ranges(
    type_bits: usize,
    total_bits: usize,
    src_bits_ident: &Ident,
    dst_bits_ident: &Ident,
    ranges: &[NormalizedRange],
) -> Result<TokenStream, String> {
    let mut statements = vec![];

    let mut dst_range_start = type_bits - total_bits;
    for range in ranges {
        statements.push(generate_bit_range_copy(
            src_bits_ident,
            dst_bits_ident,
            (range.start, range.end),
            (dst_range_start, dst_range_start + range.size()),
        ));
        dst_range_start += range.size();
    }

    Ok(quote! {{ #(#statements)* }})
}

pub fn generate_copy_to_normalized_ranges(
    type_bits: usize,
    total_bits: usize,
    src_bits_ident: &Ident,
    dst_bits_ident: &Ident,
    ranges: &[NormalizedRange],
) -> Result<TokenStream, String> {
    let mut statements = vec![];

    let mut src_range_start = type_bits - total_bits;
    for range in ranges {
        statements.push(generate_bit_range_copy(
            src_bits_ident,
            dst_bits_ident,
            (src_range_start, src_range_start + range.size()),
            (range.start, range.end),
        ));
        src_range_start += range.size();
    }

    Ok(quote! {{ #(#statements)* }})
}
