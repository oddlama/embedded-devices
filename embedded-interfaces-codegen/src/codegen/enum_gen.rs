//! Code generation for enum definitions

use proc_macro2::TokenStream as TokenStream2;
use quote::quote;

use crate::parser::{EnumDefinition, EnumPattern};

pub fn generate_enum(enum_def: &EnumDefinition) -> syn::Result<TokenStream2> {
    let enum_name = &enum_def.name;
    let underlying_type = &enum_def.underlying_type;
    let attributes = &enum_def.attributes;
    let variants = &enum_def.variants;

    // Get the effective bit size
    let bit_size = enum_def.get_effective_bit_size()?;

    // Generate enum variants
    let mut enum_variants = Vec::new();
    let mut from_unsigned_arms = Vec::new();
    let mut to_unsigned_arms = Vec::new();

    for variant in variants {
        let variant_name = &variant.name;
        let variant_attrs = &variant.attributes;

        if variant.capture_value {
            // Variant captures the underlying value
            enum_variants.push(quote! {
                #(#variant_attrs)*
                #variant_name(#underlying_type)
            });

            // For from_unsigned, we need to check if the value matches the pattern
            let pattern = generate_pattern(&variant.pattern);
            from_unsigned_arms.push(quote! {
                #pattern => Self::#variant_name(value)
            });

            // For to_unsigned, just return the captured value
            to_unsigned_arms.push(quote! {
                Self::#variant_name(v) => *v
            });
        } else {
            // Variant doesn't capture the value
            enum_variants.push(quote! {
                #(#variant_attrs)*
                #variant_name
            });

            // For from_unsigned, we need to check if the value matches the pattern
            let pattern = generate_pattern(&variant.pattern);
            from_unsigned_arms.push(quote! {
                #pattern => Self::#variant_name
            });

            // For to_unsigned, return the representative value
            let representative = &variant.representative;
            to_unsigned_arms.push(quote! {
                Self::#variant_name => #representative
            });
        }
    }

    // Generate the enum definition
    let enum_def_code = quote! {
        #(#attributes)*
        #[derive(Copy, Clone, PartialEq, Eq, core::fmt::Debug, defmt::Format)]
        pub enum #enum_name {
            #(#enum_variants,)*
        }
    };

    // Check if we have a wildcard pattern to handle unreachable case
    let has_wildcard = variants.iter().any(|v| matches!(v.pattern, EnumPattern::Wildcard));
    // Check if we use all bits in the underlying type. We only require the enum to exhaust the N
    // given bits and need to adjust the match accordingly.
    let uses_all_bits = bit_size == enum_def.get_default_bit_size()?;

    let from_unsigned_match = if has_wildcard || uses_all_bits {
        // If we have a wildcard or our match is fully exhaustive, don't add a default arm
        quote! {
            match value {
                #(#from_unsigned_arms,)*
            }
        }
    } else {
        // If there is no wildcard and our matches are not exhausting the underlying type, we need
        // to add a default case (this should not happen in practice since the trait says the
        // function must succeed for any bit value, but we'll add it for safety)
        quote! {
            match value {
                #(#from_unsigned_arms,)*
                _ => panic!("Unused bits in packed enum values must not be used! This shouldn't have happened, please file a bug report in embedded-interfaces."),
            }
        }
    };

    // Generate the UnsignedPackable implementation
    let packable_impl = quote! {
        impl embedded_interfaces::packable::UnsignedPackable for #enum_name {
            type Base = #underlying_type;
            const BITS: usize = #bit_size;

            fn from_unsigned(value: Self::Base) -> Self {
                #from_unsigned_match
            }

            fn to_unsigned(&self) -> Self::Base {
                match self {
                    #(#to_unsigned_arms,)*
                }
            }
        }
    };

    Ok(quote! {
        #enum_def_code
        #packable_impl
    })
}

/// Generate a macro pattern for a given enum pattern
fn generate_pattern(pattern: &EnumPattern) -> TokenStream2 {
    match pattern {
        EnumPattern::Single(val) => {
            quote! { #val }
        }
        EnumPattern::Range(start, end) => {
            quote! { #start..#end }
        }
        EnumPattern::RangeInclusive(start, end) => {
            quote! { #start..=#end }
        }
        EnumPattern::Multiple(items) => {
            let checks: Vec<_> = items.iter().map(generate_pattern).collect();
            quote! { #(#checks)|* }
        }
        EnumPattern::Wildcard => {
            quote! { _ }
        }
    }
}
