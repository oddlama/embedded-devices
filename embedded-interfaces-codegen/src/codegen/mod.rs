//! Code generation for register definitions

use proc_macro2::Ident;
use proc_macro2::TokenStream as TokenStream2;
use quote::{format_ident, quote};
use syn::Expr;

use crate::parser::*;

mod bit_manipulation;
mod bit_pattern;
mod packed_struct;
mod register_traits;

pub use packed_struct::*;
pub use register_traits::*;

pub fn generate_registers(registers_def: &RegistersDefinition) -> syn::Result<TokenStream2> {
    let mut generated = TokenStream2::new();

    // Generate code for each register
    for register in &registers_def.registers {
        let register_code = generate_register(registers_def, register)?;
        generated.extend(register_code);
    }

    Ok(generated)
}

fn generate_register(registers_def: &RegistersDefinition, register: &RegisterDefinition) -> syn::Result<TokenStream2> {
    let register_name = &register.name;
    let unpacked_name = format_ident!("{}Unpacked", register_name);

    // Get effective attributes (defaults + register-specific)
    let attrs = registers_def.get_effective_attrs(register)?;

    // Extract register attributes
    let (addr, mode, size) = extract_register_attrs(&register.name, &attrs)?;
    let (codec_error, i2c_codec, spi_codec) = extract_codec_attrs(&register.name, &attrs)?;

    // Generate the packed/unpacked struct pair (reusable part)
    let packed_structs = generate_packed_struct_pair(
        register_name,
        &unpacked_name,
        &register.fields,
        &register.attributes,
        size,
    )?;

    // Generate register-specific trait implementations
    let trait_impls = generate_register_trait_implementations(
        register_name,
        &unpacked_name,
        &addr,
        &mode,
        size,
        &codec_error,
        &i2c_codec,
        &spi_codec,
    )?;

    Ok(quote! {
        #packed_structs
        #trait_impls
    })
}

fn extract_register_attrs(register_name: &Ident, attrs: &[Attr]) -> syn::Result<(TokenStream2, String, usize)> {
    let mut addr = None;
    let mut mode = None;
    let mut size = None;

    for attr in attrs {
        match attr.name.to_string().as_str() {
            "addr" => {
                let attr_value = &attr.value;
                addr = Some(quote! { #attr_value });
            }
            "mode" => {
                // Extract mode string from expression
                mode = Some(extract_mode_string(&attr.value)?);
            }
            "size" => {
                // Extract size integer from expression
                size = Some(extract_size_integer(&attr.value)?);
            }
            _ => {}
        }
    }

    let addr = addr.ok_or_else(|| syn::Error::new_spanned(register_name, "Missing required 'addr' attribute"))?;
    let mode = mode.ok_or_else(|| syn::Error::new_spanned(register_name, "Missing required 'mode' attribute"))?;
    let size = size.ok_or_else(|| syn::Error::new_spanned(register_name, "Missing required 'size' attribute"))?;

    Ok((addr, mode, size))
}

fn extract_codec_attrs(
    register_name: &Ident,
    attrs: &[Attr],
) -> syn::Result<(TokenStream2, TokenStream2, TokenStream2)> {
    let mut codec_error = None;
    let mut i2c_codec = None;
    let mut spi_codec = None;

    for attr in attrs {
        let attr_value = &attr.value;
        match attr.name.to_string().as_str() {
            "codec_error" => {
                codec_error = Some(quote! { #attr_value });
            }
            "i2c_codec" => {
                i2c_codec = Some(quote! { #attr_value });
            }
            "spi_codec" => {
                spi_codec = Some(quote! { #attr_value });
            }
            _ => {}
        }
    }

    let codec_error = codec_error
        .ok_or_else(|| syn::Error::new_spanned(register_name, "Missing required 'codec_error' attribute"))?;
    let i2c_codec =
        i2c_codec.ok_or_else(|| syn::Error::new_spanned(register_name, "Missing required 'i2c_codec' attribute"))?;
    let spi_codec =
        spi_codec.ok_or_else(|| syn::Error::new_spanned(register_name, "Missing required 'spi_codec' attribute"))?;

    Ok((codec_error, i2c_codec, spi_codec))
}

/// Extract a mode string from an expression
fn extract_mode_string(expr: &Expr) -> syn::Result<String> {
    match expr {
        Expr::Lit(syn::ExprLit {
            lit: syn::Lit::Str(lit_str),
            ..
        }) => Ok(lit_str.value()),
        Expr::Path(path) => {
            if let Some(ident) = path.path.get_ident() {
                Ok(ident.to_string())
            } else {
                Err(syn::Error::new_spanned(
                    expr,
                    "Mode must be a string literal or identifier (e.g., \"rw\", r, rw)",
                ))
            }
        }
        _ => Err(syn::Error::new_spanned(
            expr,
            "Mode must be a string literal or identifier (e.g., \"rw\", r, rw)",
        )),
    }
}

/// Extract a size integer from an expression
fn extract_size_integer(expr: &Expr) -> syn::Result<usize> {
    match expr {
        Expr::Lit(syn::ExprLit {
            lit: syn::Lit::Int(lit_int),
            ..
        }) => lit_int.base10_parse::<usize>(),
        _ => Err(syn::Error::new_spanned(expr, "Size must be an integer literal")),
    }
}
