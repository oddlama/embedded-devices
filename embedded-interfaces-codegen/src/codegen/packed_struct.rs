//! Reusable packed/unpacked struct generation
//!
//! This module contains the generic logic for generating packed and unpacked
//! struct pairs that can be reused for registers and other similar structures.

use proc_macro2::TokenStream as TokenStream2;
use quote::quote;
use syn::{Attribute, Ident};

use super::bit_pattern::{generate_bit_pattern_doc, process_field_bit_patterns};
use crate::parser::FieldDefinition;

/// Generate a pair of packed and unpacked structs with conversion implementations
///
/// This function is generic and can be used for registers or any other structures
/// that need packed/unpacked representations.
pub fn generate_packed_struct_pair(
    packed_name: &Ident,
    unpacked_name: &Ident,
    fields: &[FieldDefinition],
    doc_attrs: &[Attribute],
    size: usize,
) -> syn::Result<TokenStream2> {
    // Process and validate bit patterns
    let total_size_bits = (size * 8) as u32;
    let processed_fields = process_field_bit_patterns(packed_name, fields, total_size_bits)?;

    let unpacked_struct = generate_unpacked_struct(&processed_fields, packed_name, unpacked_name, doc_attrs, size)?; // Pass size
    let packed_struct = generate_packed_struct(&processed_fields, packed_name, unpacked_name, size, doc_attrs)?;
    let conversions = generate_struct_conversions(packed_name, unpacked_name)?;

    Ok(quote! {
        #unpacked_struct
        #packed_struct
        #conversions
    })
}

fn generate_unpacked_struct(
    processed_fields: &[super::bit_pattern::ProcessedField],
    packed_name: &Ident,
    unpacked_name: &Ident,
    doc_attrs: &[Attribute],
    size: usize, // Add size parameter
) -> syn::Result<TokenStream2> {
    let mut struct_fields = Vec::new();
    let mut default_values = Vec::new();

    for processed in processed_fields {
        let field = &processed.field;
        // Skip reserved fields
        if field.is_reserved() {
            continue;
        }

        let field_name = &field.name;
        let field_type = &field.field_type;
        let field_attrs = &field.attributes;

        // Generate field documentation
        let default_doc = if let Some(default_val) = &field.default_value {
            quote! { #[doc = concat!("Default: `", stringify!(#default_val), "`")] }
        } else {
            quote! { #[doc = "Default: not specified"] }
        };

        // Generate bit pattern documentation
        let bit_pattern_doc = generate_bit_pattern_doc(&processed.normalized_ranges);
        let bit_doc = quote! { #[doc = #bit_pattern_doc] };

        struct_fields.push(quote! {
            #(#field_attrs)*
            #[doc = ""]
            #default_doc
            #bit_doc
            pub #field_name: #field_type
        });

        // Collect default values
        if let Some(default_val) = &field.default_value {
            default_values.push(quote! { #field_name: #default_val });
        } else {
            default_values.push(quote! { #field_name: Default::default() });
        }
    }

    // Generate the byte conversion method
    let pack_body = super::pack::generate_pack_body(packed_name, processed_fields, size)?;

    Ok(quote! {
        #(#doc_attrs)*
        #[derive(Copy, Clone, PartialEq, core::fmt::Debug, defmt::Format)]
        pub struct #unpacked_name {
            #(#struct_fields,)*
        }

        impl #unpacked_name {
            #[inline]
            #[doc = concat!("Pack all fields into a [`", stringify!(#packed_name), "`].")]
            pub fn pack(&self) -> #packed_name {
                #pack_body
            }
        }

        impl Default for #unpacked_name {
            fn default() -> Self {
                Self {
                    #(#default_values,)*
                }
            }
        }
    })
}

/// Generate the packed struct with byte array representation
fn generate_packed_struct(
    processed_fields: &[super::bit_pattern::ProcessedField],
    packed_name: &Ident,
    unpacked_name: &Ident,
    size: usize,
    doc_attrs: &[Attribute],
) -> syn::Result<TokenStream2> {
    // Generate the byte conversion method
    let unpack_body = super::unpack::generate_unpack_body(unpacked_name, processed_fields)?;

    Ok(quote! {
        #(#doc_attrs)*
        #[doc = concat!("This is the packed representation of [`", stringify!(#unpacked_name), "`].")]
        #[derive(Copy, Clone, PartialEq, Eq, bytemuck::Pod, bytemuck::Zeroable)]
        #[repr(transparent)]
        pub struct #packed_name([u8; #size]);

        impl #packed_name {
            #[inline]
            #[doc = concat!("Unpack all fields into a [`", stringify!(#unpacked_name), "`]. If you don't need all fields, this is more expensive than just using the appropriate `read_*` functions directly.")]
            pub fn unpack(&self) -> #unpacked_name {
                #unpack_body
            }
        }

        impl Default for #packed_name {
            fn default() -> Self {
                #unpacked_name::default().pack()
            }
        }
    })
}

/// Generate conversion implementations between packed and unpacked structs
fn generate_struct_conversions(packed_name: &Ident, unpacked_name: &Ident) -> syn::Result<TokenStream2> {
    Ok(quote! {
        impl AsRef<#unpacked_name> for #unpacked_name {
            #[inline]
            fn as_ref(&self) -> &#unpacked_name {
                self
            }
        }

        impl AsRef<#packed_name> for #packed_name {
            #[inline]
            fn as_ref(&self) -> &#packed_name {
                self
            }
        }

        impl From<&#packed_name> for #unpacked_name {
            #[inline]
            fn from(value: &#packed_name) -> Self {
                value.unpack()
            }
        }

        impl From<#packed_name> for #unpacked_name {
            #[inline]
            fn from(value: #packed_name) -> Self {
                value.unpack()
            }
        }

        impl From<&#unpacked_name> for #packed_name {
            #[inline]
            fn from(value: &#unpacked_name) -> Self {
                value.pack()
            }
        }

        impl From<#unpacked_name> for #packed_name {
            #[inline]
            fn from(value: #unpacked_name) -> Self {
                value.pack()
            }
        }
    })
}
