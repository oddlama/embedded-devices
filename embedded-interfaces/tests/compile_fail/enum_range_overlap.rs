use embedded_interfaces::codegen::interface_objects;
interface_objects! {
    enum Overlap: u8{4} {
        0..=5 Range1,
        4..=8 Range2, // 4-5 overlap
        _ Wild,
    }
}
fn main() {}
