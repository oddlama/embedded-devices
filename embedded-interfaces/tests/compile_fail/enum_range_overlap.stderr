error: Value 4 is captured by multiple variants
 --> tests/compile_fail/enum_range_overlap.rs:5:15
  |
5 |         4..=8 Range2, // 4-5 overlap
  |               ^^^^^^
